src/tablespace.c

```diff
@@ -0,0 +1,467 @@
+{
+	Catalog    *catalog = catalog_get();
+	Relation	rel;
+	int32		id;
+
+	rel = heap_open(catalog->tables[TABLESPACE].id, RowExclusiveLock);
+	id = tablespace_insert_relation(rel, hypertable_id, tspcname);
+	heap_close(rel, RowExclusiveLock);
+
+	return id;
+}
+
+typedef struct TablespaceScanInfo
+{
+	Catalog    *catalog;
+	Cache	   *hcache;
+	Oid			userid;
+	bool		stopcount;
+} TablespaceScanInfo;
+
+static bool
+tablespace_tuple_delete(TupleInfo *ti, void *data)
+{
+	TablespaceScanInfo *info = data;
+	CatalogSecurityContext sec_ctx;
+
+	catalog_become_owner(info->catalog, &sec_ctx);
+	catalog_delete_only(ti->scanrel, ti->tuple);
+	catalog_restore_user(&sec_ctx);
+
+	return (info->stopcount == 0 || ti->count < info->stopcount);
+}
+
+static int
+tablespace_delete(int32 hypertable_id, const char *tspcname)
+{
+	ScanKeyData scankey[2];
+	TablespaceScanInfo info = {
+		.catalog = catalog_get(),
+		.stopcount = (NULL != tspcname),
+	};
+	ScannerCtx	scanctx = {
+		.table = info.catalog->tables[TABLESPACE].id,
+		.index = info.catalog->tables[TABLESPACE].index_ids[TABLESPACE_HYPERTABLE_ID_TABLESPACE_NAME_IDX],
+		.scantype = ScannerTypeIndex,
+		.nkeys = 0,
+		.scankey = scankey,
+		.tuple_found = tablespace_tuple_delete,
+		.data = &info,
+		.lockmode = RowExclusiveLock,
+		.scandirection = ForwardScanDirection,
+	};
+	int			num_deleted;
+
+	ScanKeyInit(&scankey[scanctx.nkeys++],
+			 Anum_tablespace_hypertable_id_tablespace_name_idx_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(hypertable_id));
+
+	if (NULL != tspcname)
+		ScanKeyInit(&scankey[scanctx.nkeys++],
+		   Anum_tablespace_hypertable_id_tablespace_name_idx_tablespace_name,
+					BTEqualStrategyNumber,
+					F_NAMEEQ,
+					DirectFunctionCall1(namein, CStringGetDatum(tspcname)));
+
+	num_deleted = scanner_scan(&scanctx);
+
+	if (num_deleted > 0)
+	{
+		catalog_invalidate_cache(catalog_table_get_id(info.catalog, TABLESPACE), CMD_DELETE);
+		CommandCounterIncrement();
+	}
+
+	return num_deleted;
+}
+
+static bool
+tablespace_tuple_owner_filter(TupleInfo *ti, void *data)
+{
+	TablespaceScanInfo *info = data;
+	FormData_tablespace *form = (FormData_tablespace *) GETSTRUCT(ti->tuple);
+	Hypertable *ht;
+
+	ht = hypertable_cache_get_entry_by_id(info->hcache, form->hypertable_id);
+
+	if (ht == NULL)
+		elog(ERROR, "No hypertable exists for tablespace \"%s\"",
+			 NameStr(form->tablespace_name));
+
+	return hypertable_has_privs_of(ht->main_table_relid, info->userid);
+}
+
+static int
+tablespace_delete_from_all(const char *tspcname, Oid userid)
+{
+	ScanKeyData scankey[1];
+	TablespaceScanInfo info = {
+		.catalog = catalog_get(),
+		.hcache = hypertable_cache_pin(),
+		.userid = userid,
+	};
+	ScannerCtx	scanctx = {
+		.table = info.catalog->tables[TABLESPACE].id,
+		.scantype = ScannerTypeHeap,
+		.nkeys = 1,
+		.scankey = scankey,
+		.tuple_found = tablespace_tuple_delete,
+		.filter = tablespace_tuple_owner_filter,
```

Comment:
It is not a failure so why should we warn? It is just a scan for tablespace mappings where you have owner permission on the associated table.

Reply:
Besides, warning would just print a lot of warnings about tablespaces and tables you have nothing to do with, for no good reason I can see.

Reply:
IMHO, I think it is fine that it is implicit that this applies only to your own hypertables. The fact that we store all hypertables in the same internal metadata table rather than segmented on user is not something that I believe we should expose to the user. Would you feel the same thing if we'd have a column in the tablespace table with `userId` and do a scan with `current_user = userId`? Would that also require printing warnings for  the case when `userId != current_user`?

If you really feel strongly about this I guess we could print a single NOTICE at the end that the tablespace remains attached to some hypertables if that's the case?

Maybe others can chime in?

Reply:
I think I agree there should be a NOTICE; a single one at the end is probably best

Reply:
Yeah I would feel the same if userId was a column. I think this is a semantics things. Will ask others

Reply:
Added a NOTICE at the end if that works for you.

Reply:
LGTM